### DISCUSSION TITLE
Transparent BackgroundColor turns black

### DISCUSSION THREAD

**Discussion by tryashtar** [SixLabors/ImageSharp#2399] [Q&A] [answered]
Hello :) I am trying to save an image with a _white_ transparent background.

This is what I tried:

```cs
using var test = Image.Load<Rgba32>("in.png");
test.Mutate(x => { x.BackgroundColor(Color.FromRgba(255, 255, 255, 0)); });
var encoder = new PngEncoder() { TransparentColorMode = PngTransparentColorMode.Preserve };
test.SaveAsPng("out.png", encoder);
```

Using this image:
![in](https://user-images.githubusercontent.com/26075577/224611764-75953f50-1f71-44d3-a45e-f616d34a9d5e.png)

Unfortunately, the result turns out unchanged: it has a _black_ transparent background. I tried messing around with different `GraphicsOptions`, but nothing seemed to do the trick. The only solution I've found is setting the alpha to 1 instead of 0, but that doesn't feel right :(

Is this possible? Did I miss something easy? Thanks for your work and help!

**Comment by JimBobSquarePants** [comment, marked as answer]
I cannot replicate this. 

<img  alt="image" src="https://user-images.githubusercontent.com/385879/224620734-1b54fb73-3db3-4b05-b1e4-2f0240d5ede9.png">


<img alt="image" src="https://user-images.githubusercontent.com/385879/224620802-c4ca01eb-5b37-4fda-a5c9-3ec6edc4575b.png">

<img  alt="image" src="https://user-images.githubusercontent.com/385879/224620867-f31da298-09c8-4e0b-bc8f-b9f8b9652c57.png">

**Reply by tryashtar** [reply]
So for you it looks like this
![image](https://user-images.githubusercontent.com/26075577/224625460-baa31fbd-af84-4908-82d5-720ff1186853.png)
instead of this?
![image](https://user-images.githubusercontent.com/26075577/224625502-e1d6e4be-dd71-4e17-8af2-096adad90e7d.png)

I'm very confused then. I tried even on different platforms and got the second for both...

**Reply by JimBobSquarePants** [reply]
Ah... I read that at black, not black transparent (There's no such thing technically as black transparent btw).

That result is expected. 

Replacing the background requires an operation called alpha composition. In order to do alpha composition, you need to premultiply the color RGB components by their A component. 

(255, 255, 255) x 0 = (0, 0, 0)

**Reply by tryashtar** [reply]
Yeah, black transparency. Sorry if that wasn't clear in my post. It's a bit of a weird thing to want, so I was glad to see the `PngTransparentColorMode.Preserve` option. But I'm not sure why it's not doing what I expect here

**Reply by JimBobSquarePants** [reply]
Preserve is too late. The color has already been converted to (0, 0, 0, 0) due to the premultiplication.

**Reply by tryashtar** [reply]
Mm, I see. I guess I can always just `ProcessPixelRows` to set them how I want, if there's no easier way. Thanks for looking!

**Reply by JimBobSquarePants** [reply]
No worries. 

Here's possibly the best article I've ever read on alpha compositing. 
https://ciechanow.ski/alpha-compositing/

About half-way down there an interactive image which demonstrates exactly what happens when the source color alpha is 0.

<img alt="image" src="https://user-images.githubusercontent.com/385879/224632366-d0a90a88-b228-45a2-93b1-80f3dac06893.png">
